Add more detailed documentation on how to use multiple inventories

zharalim opened the pull request:
##### SUMMARY

Add a separate section on how to use multiple inventory directories with an example.

~~The documentation on multiple inventories was a bit hard to find as it was located only in the dynamic inventory documentation page. In addition that section doesn't have many details on how multiple inventories actually work.~~

This seems to be an undocumented feature related to "multiple sources" described in *Working with dynamic inventory*. Multiple sources documentation talks about using a directory and having multiple different inventory sources inside the directory. The new documented use case is different as it uses multiple inventory directories to promote DRY principle. The new documentation is in line with the comments in #36600 by @bcoca.

Using multiple directories is a bit cumbersome because you need to use multiple `-i` parameters but with `ansible.cfg` or `ANSIBLE_INVENTORY` it's manageable. 

I understand if this is not a feature you want to promote in the docs, it resolved some issues for me so I thought that I'll try documenting it. 

Comments appreciated :)

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
documentation

##### ANSIBLE VERSION
Tested the example with 2.7. Should also work at least with 2.6.
```
ansible 2.7.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
N/A

zharalim committed:
Add a new section on how to use multiple inventory sources w/ an example

zharalim committed:
Change "multiple inventory sources" to "multiple inventory directories"

zharalim commented:
I had another look on the dynamic inventory docs and realized that "Using Inventory Directories and Multiple Inventory Sources" is a related feature but not exactly the same thing. I removed the reference to that section. I also updated the pr description to better describe the issue.

bcoca commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -478,12 +478,12 @@ Here is an example of how to instantly deploy to created containers::
+++++++++++++++++++++++++++++++++++++
 
-You can use multiple inventory sources if you want to target multiple environments or reuse common
+You can use multiple inventory directories if you want to target multiple environments or reuse common
```
this is not really limited to directories, the original sources was correct

zharalim replied:
Fixed

bcoca reviewed (changes requested):
You might want to start with 'multiple sources' and then show that using a directory is a way to aggregate sources and then give examples on that.

zharalim committed:
Add an example on aggregating inv. sources with a directory.

zharalim commented:
I added an example on how to aggregate sources with a directory and tried to take different inventory sources into consideration better than previously.

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -14,7 +14,7 @@ You can specify a different inventory file using the ``-i <path>`` option on the
 Not only is this inventory configurable, but you can also use multiple inventory files at the same time and
 pull inventory from dynamic or cloud sources or different formats (YAML, ini, etc), as described in :ref:`intro_dynamic_inventory`.
-Introduced in version 2.4, Ansible has inventory plugins to make this flexible and customizable.
+Introduced in version 2.4, Ansible has :ref:`inventory_plugins` to make this flexible and customizable.
```
+1 for adding more links!

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -478,6 +478,96 @@ Here is an example of how to instantly deploy to created containers::
 
+.. _using_multiple_inventory_sources:
+
+Using multiple inventory sources
```
What is the use case here? Why would I want to structure my static inventory this way? What problem does this structure solve? If there's a specific scenario, that makes the example easier to follow.

bcoca replied:
so you can have specific inventories for different things, but then combine them when you need 'global access' i.e `-i /path/to/development -i /otherpath/to/production`

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -478,6 +478,96 @@ Here is an example of how to instantly deploy to created containers::
+db1.example.com will have host variables from both db-servers.yml and all-servers.yml.
+In case of conflict a variable defined in db-servers.yml wins over all-servers.yml's variable.
+
+Without the empty web-servers and db-servers groups the common inventory would have a dependency on those
```
Ah, this looks like it might define the use case. Can you add variables to the example, and then flesh this part out? Something like, "Without the empty web-servers and db-servers groups, the value of `my_var` would be X instead of Y" so it's easier to see the consequences (and advantages).

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -478,6 +478,96 @@ Here is an example of how to instantly deploy to created containers::
+        all.yml                 # assign variables to all hosts
+
+The inventories are parsed in alphabetical order so number prefixes are used to make sure
+the parsing order is correct. For more details about inventory plugins see :ref:`inventory_plugins`.
```
Why would I want to control the parsing order in this particular way? What happens if I get this wrong?

bcoca replied:
parsing order can control conflicts with variables, if you define a variable in a group twice with different values, last update wins. 

i.e  in the `[all:vars]`  in01_stuff.ini defines `myvar=1`, but 03_stuff.ini defines `myvar=2`, so in the end `myvar == 2`

acozine reviewed (commented):
This is great information - thank you @zharalim for putting it together. I think it could be made even better by making the example more detailed, to illustrate what happens when you amalgamate inventory sources in this way. Are you interested in working on this more?

zharalim commented:
I'll see if I can make the use cases clearer and improve the examples.

zharalim committed:
Simpler examples and use cases. Remove common inventory example.

zharalim commented:
Okay, hope it's a bit better now :)

Summary:
- Simplified the examples
- Described the use cases better and added examples on how variable conflicts work

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+Keep in mind that if there are variable conflicts in the inventories, they are resolved according
+to the rules described in :ref:`how_we_merge` and :ref:`ansible_variable_precedence`.
+The merging order is controlled by the order of the inventory source parameters.
+If `[all:vars]` in staging inventory defines `myvar = 1`, but production inventory defines `myvar = 2`,
```
This is a great description - very clear and practical.

zharalim replied:
Thanks!

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+================================
+
+As an advanced use case you can target multiple inventory sources (directories, dynamic inventory scripts 
+or files supported by inventory plugins) at same time by giving multiple inventory parameters from command 
```
```suggestion
or files supported by inventory plugins) at the same time by giving multiple inventory parameters from the command 
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+**Aggregating inventory sources with a directory**
+
+You can also create an inventory by combining multiple inventory sources and source types under a directory.
+This can be useful for combining static and dynamic hosts and manage them as one inventory.
```
```suggestion
This can be useful for combining static and dynamic hosts and managing them as one inventory.
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+It can be useful to control the merging order of the inventory sources if there's variable
+conflicts or group of groups dependencies to the other inventory sources. The inventories
+are merged in alphabetical order according to the filenames so the result can
+be controller by adding prefixes to the files::
```
```suggestion
be controlled by adding prefixes to the files::
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+Keep in mind that if there are variable conflicts in the inventories, they are resolved according
+to the rules described in :ref:`how_we_merge` and :ref:`ansible_variable_precedence`.
+The merging order is controlled by the order of the inventory source parameters.
+If `[all:vars]` in staging inventory defines `myvar = 1`, but production inventory defines `myvar = 2`,
```
```suggestion
If ``[all:vars]`` in staging inventory defines ``myvar = 1``, but production inventory defines ``myvar = 2``,
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+to the rules described in :ref:`how_we_merge` and :ref:`ansible_variable_precedence`.
+The merging order is controlled by the order of the inventory source parameters.
+If `[all:vars]` in staging inventory defines `myvar = 1`, but production inventory defines `myvar = 2`,
+the playbook will be run with `myvar = 2`. The result would be reversed if the playbook was run with
```
```suggestion
the playbook will be run with ``myvar = 2``. The result would be reversed if the playbook was run with
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+The merging order is controlled by the order of the inventory source parameters.
+If `[all:vars]` in staging inventory defines `myvar = 1`, but production inventory defines `myvar = 2`,
+the playbook will be run with `myvar = 2`. The result would be reversed if the playbook was run with
+`-i production -i staging`.
```
```suggestion
``-i production -i staging``.
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+      group_vars/
+        all.yml                 # assign variables to all hosts
+
+If the openstack inventory source defines `myvar = 1` for the group `all`, dynamic inventory source defines `myvar = 2`,
```
```suggestion
If ``01-openstack.yml`` defines ``myvar = 1`` for the group ``all``, ``02-dynamic-inventory.py`` defines ``myvar = 2``,
```

acozine commented on docs/docsite/rst/user_guide/intro_inventory.rst:
```diff
@@ -354,6 +354,63 @@ Starting in Ansible version 2.4, users can use the group variable ``ansible_grou
+        all.yml                 # assign variables to all hosts
+
+If the openstack inventory source defines `myvar = 1` for the group `all`, dynamic inventory source defines `myvar = 2`,
+and `03-static-inventory` defines `myvar = 3`, the playbook will be run with `myvar = 3`.
```
```suggestion
and ``03-static-inventory`` defines ``myvar = 3``, the playbook will be run with ``myvar = 3``.
```

acozine reviewed (commented):
Thanks @zharalim this looks great. I made a few suggestions - corrected a couple of typos and changed the single backticks (which publish as _italics_ in Sphinx) to double backticks (which publish as `code` in Sphinx). If that's not what you intended, let me know. It's annoying that markdown and rST/Sphinx do that differently - I make that mistake a lot.

acozine committed:
Fix code blocks and typos

Co-Authored-By: zharalim <zharalim@outlook.com>

zharalim commented:
I committed the suggestions. Thanks for the feedback :)

acozine commented:
Thanks @zharalim for making the Ansible docs more comprehensive!
